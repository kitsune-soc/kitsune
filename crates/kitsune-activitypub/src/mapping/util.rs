use super::State;
use kitsune_db::model::{account::Account, post::Visibility};
use kitsune_type::ap::PUBLIC_IDENTIFIER;

pub trait BaseToCc {
    fn base_to_cc(&self, state: State<'_>, account: &Account) -> (Vec<String>, Vec<String>);
}

impl BaseToCc for Visibility {
    #[inline]
    fn base_to_cc(&self, state: State<'_>, account: &Account) -> (Vec<String>, Vec<String>) {
        let followers_url = state.service.url.followers_url(account.id);

        match self {
            Visibility::Public => (vec![PUBLIC_IDENTIFIER.to_string()], vec![followers_url]),
            Visibility::Unlisted => (vec![], vec![PUBLIC_IDENTIFIER.to_string(), followers_url]),
            Visibility::FollowerOnly => (vec![followers_url], vec![]),
            Visibility::MentionOnly => (vec![], vec![]),
        }
    }
}
