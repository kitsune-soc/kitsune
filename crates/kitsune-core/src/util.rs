use crate::state::State;
use hex_simd::{AsOut, AsciiCase};
use iso8601_timestamp::Timestamp;
use kitsune_db::model::{account::Account, oauth2::access_token::AccessToken, post::Visibility};
use kitsune_type::ap::PUBLIC_IDENTIFIER;
use pulldown_cmark::{html, Options, Parser};

const TOKEN_LENGTH: usize = 32;

#[inline]
#[must_use]
pub fn generate_secret() -> String {
    let token_data: [u8; TOKEN_LENGTH] = rand::random();
    let mut buf = [0_u8; TOKEN_LENGTH * 2];
    (*hex_simd::encode_as_str(&token_data, buf.as_mut_slice().as_out(), AsciiCase::Lower))
        .to_string()
}

#[inline]
#[must_use]
pub fn process_markdown(markdown: &str) -> String {
    let parser = Parser::new_ext(markdown, Options::all());
    let mut buf = String::new();
    html::push_html(&mut buf, parser);
    buf
}

pub trait AccessTokenTtl {
    fn ttl(&self) -> time::Duration;
}

impl AccessTokenTtl for AccessToken {
    #[inline]
    fn ttl(&self) -> time::Duration {
        *self.expires_at - *Timestamp::now_utc()
    }
}

pub trait BaseToCc {
    fn base_to_cc(&self, state: &State, account: &Account) -> (Vec<String>, Vec<String>);
}

impl BaseToCc for Visibility {
    #[inline]
    fn base_to_cc(&self, state: &State, account: &Account) -> (Vec<String>, Vec<String>) {
        let followers_url = state.service.url.followers_url(account.id);

        match self {
            Visibility::Public => (vec![PUBLIC_IDENTIFIER.to_string()], vec![followers_url]),
            Visibility::Unlisted => (vec![], vec![PUBLIC_IDENTIFIER.to_string(), followers_url]),
            Visibility::FollowerOnly => (vec![followers_url], vec![]),
            Visibility::MentionOnly => (vec![], vec![]),
        }
    }
}
