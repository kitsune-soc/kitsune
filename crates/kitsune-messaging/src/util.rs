use std::fmt;

pub struct TransparentDebug<T>(pub T);

impl<T> fmt::Debug for TransparentDebug<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{} {{ ... }}", std::any::type_name::<T>())
    }
}
