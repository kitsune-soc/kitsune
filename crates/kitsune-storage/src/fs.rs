//!
//! File system backed implementation of the [`StorageBackend`] trait
//!

use crate::{Result, StorageBackend};
use async_trait::async_trait;
use bytes::Bytes;
use futures_util::{pin_mut, stream::BoxStream, Stream, StreamExt, TryStreamExt};
use std::path::PathBuf;
use tokio::{
    fs::{self, File},
    io::AsyncWriteExt,
};
use tokio_util::io::ReaderStream;

#[derive(Clone)]
/// File system storage
pub struct Storage {
    storage_dir: PathBuf,
}

impl Storage {
    /// Create a new file system storage
    ///
    /// It always requires a directory the operations on it are relative to
    #[must_use]
    pub fn new(storage_dir: PathBuf) -> Self {
        Self { storage_dir }
    }
}

#[async_trait]
impl StorageBackend for Storage {
    async fn delete(&self, path: &str) -> Result<()> {
        fs::remove_file(self.storage_dir.join(path)).await?;
        Ok(())
    }

    async fn get(&self, path: &str) -> Result<BoxStream<'static, Result<Bytes>>> {
        let file = File::open(self.storage_dir.join(path)).await?;
        Ok(ReaderStream::new(file).map_err(Into::into).boxed())
    }

    async fn put<T>(&self, path: &str, input_stream: T) -> Result<()>
    where
        T: Stream<Item = Result<Bytes>> + Send + 'static,
    {
        let mut file = File::create(self.storage_dir.join(path)).await?;
        pin_mut!(input_stream);
        while let Some(chunk) = input_stream.next().await.transpose()? {
            file.write_all(&chunk).await?;
        }
        file.flush().await?;

        Ok(())
    }
}

#[cfg(test)]
mod test {
    use crate::{fs::Storage, StorageBackend};
    use bytes::{BufMut, BytesMut};
    use futures_util::{future, stream, StreamExt, TryStreamExt};
    use std::str;
    use tempfile::TempDir;

    const TEST_TEXT: &str = r#"
    新時代はこの未来だ
    世界中全部 変えてしまえば 変えてしまえば
    ジャマモノ やなもの なんて消して
    この世とメタモルフォーゼしようぜ
    ミュージック キミが起こす マジック
    目を閉じれば未来が開いて
    いつまでも終わりが来ないようにって
    この歌を歌うよ
    Do you wanna play? リアルゲーム ギリギリ
    綱渡りみたいな旋律 認めない戻れない忘れたい
    夢の中に居させて I wanna be free
    見えるよ新時代が 世界の向こうへ
    さあ行くよ new world
    新時代はこの未来だ
    世界中全部 変えてしまえば 変えてしまえば
    果てしない音楽がもっと届くように
    夢は見ないわ キミが話した「ボクを信じて」
    Ooh
    あれこれいらないものは消して
    リアルをカラフルに越えようぜ
    ミュージック 今始まる ライジング
    目をつぶりみんなで逃げようよ
    今よりイイモノを見せてあげるよ
    この歌を歌えば
    Do you wanna play? リアルゲーム ギリギリ
    綱渡りみたいな運命 認めない戻れない忘れたい
    夢の中に居させて I wanna be free
    見えるよ新時代が 世界の向こうへ
    さあ行くよ new world
    信じたいわ この未来を
    世界中全部 変えてしまえば 変えてしまえば
    果てしない音楽がもっと届くように
    夢を見せるよ 夢を見せるよ 新時代だ
    Ooh
    新時代だ"#;

    #[tokio::test]
    async fn basic() {
        let temp_dir = TempDir::new().unwrap();
        let storage = Storage::new(temp_dir.path().to_owned());

        storage
            .put(
                "hello-world",
                stream::once(future::ok(TEST_TEXT.into())).boxed(),
            )
            .await
            .unwrap();

        let data = storage
            .get("hello-world")
            .await
            .unwrap()
            .try_fold(BytesMut::new(), |mut acc, chunk| {
                acc.put(chunk);
                future::ok(acc)
            })
            .await
            .unwrap();
        let data = str::from_utf8(&data).unwrap();

        assert_eq!(TEST_TEXT, data);

        storage.delete("hello-world").await.unwrap();
        assert!(storage.get("hello-world").await.is_err());
    }
}
