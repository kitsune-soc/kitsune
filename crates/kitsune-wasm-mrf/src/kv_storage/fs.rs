use super::{Backend, BoxError, BucketBackend};
use miette::IntoDiagnostic;
use std::path::Path;

pub struct FsBackend {
    inner: sled::Db,
}

impl FsBackend {
    pub fn from_path<P>(path: P) -> miette::Result<Self>
    where
        P: AsRef<Path>,
    {
        Ok(Self {
            inner: sled::open(path).into_diagnostic()?,
        })
    }
}

impl Backend for FsBackend {
    type Bucket = FsBucketBackend;

    async fn open(&self, module_name: &str, name: &str) -> Result<Self::Bucket, BoxError> {
        self.inner
            .open_tree(format!("{module_name}:{name}"))
            .map(|tree| FsBucketBackend { inner: tree })
            .map_err(Into::into)
    }
}

pub struct FsBucketBackend {
    inner: sled::Tree,
}

impl BucketBackend for FsBucketBackend {
    async fn exists(&self, key: &str) -> Result<bool, BoxError> {
        self.inner.contains_key(key).map_err(Into::into)
    }

    async fn delete(&self, key: &str) -> Result<(), BoxError> {
        self.inner.remove(key)?;
        Ok(())
    }

    async fn get(&self, key: &str) -> Result<Option<Vec<u8>>, BoxError> {
        self.inner
            .get(key)
            .map(|maybe_val| maybe_val.map(|val| val.to_vec()))
            .map_err(Into::into)
    }

    async fn set(&self, key: &str, value: &[u8]) -> Result<(), BoxError> {
        self.inner.insert(key, value)?;
        Ok(())
    }
}
