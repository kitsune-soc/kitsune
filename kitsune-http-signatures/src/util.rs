use std::time::{Duration, SystemTime, SystemTimeError};

pub trait UnixTimestampExt {
    type Error;

    fn from_unix_timestamp(timestamp: u64) -> Self;
    fn to_unix_timestamp(&self) -> Result<u64, Self::Error>;
}

impl UnixTimestampExt for SystemTime {
    type Error = SystemTimeError;

    fn from_unix_timestamp(timestamp: u64) -> Self {
        let duration = Duration::from_secs(timestamp);
        SystemTime::UNIX_EPOCH + duration
    }

    fn to_unix_timestamp(&self) -> Result<u64, Self::Error> {
        self.duration_since(SystemTime::UNIX_EPOCH)
            .map(|duration| duration.as_secs())
    }
}
