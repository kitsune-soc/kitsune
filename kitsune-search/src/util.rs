//!
//! Utility functions
//!

use std::ops::Bound;

/// Extension trait on [`Bound`]
pub trait BoundExt<T> {
    /// Vendored until [rust-lang/rust#86026](https://github.com/rust-lang/rust/issues/86026) is stabilised
    fn map<U, F: FnOnce(T) -> U>(self, f: F) -> Bound<U>;
}

impl<T> BoundExt<T> for Bound<T> {
    #[inline]
    fn map<U, F: FnOnce(T) -> U>(self, f: F) -> Bound<U> {
        match self {
            Bound::Unbounded => Bound::Unbounded,
            Bound::Included(x) => Bound::Included(f(x)),
            Bound::Excluded(x) => Bound::Excluded(f(x)),
        }
    }
}

/// Vendored until [rust-lang/rust#88581](https://github.com/rust-lang/rust/issues/88581) is stabilised
#[inline]
#[must_use]
pub const fn div_ceil(lhs: usize, rhs: usize) -> usize {
    let d = lhs / rhs;
    let r = lhs % rhs;
    if r > 0 && rhs > 0 {
        d + 1
    } else {
        d
    }
}
