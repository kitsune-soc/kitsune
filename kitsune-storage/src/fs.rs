//!
//! File system backed implementation of the [`StorageBackend`] trait
//!

use crate::{Result, StorageBackend};
use async_trait::async_trait;
use bytes::Bytes;
use futures_util::{stream::BoxStream, Stream, StreamExt, TryStreamExt};
use std::path::PathBuf;
use tokio::{
    fs::{self, File},
    io::AsyncWriteExt,
};
use tokio_util::io::ReaderStream;

/// File system storage
pub struct Storage {
    storage_dir: PathBuf,
}

impl Storage {
    /// Create a new file system storage
    ///
    /// It always requires a directory the operations on it are relative to
    #[must_use]
    pub fn new(storage_dir: PathBuf) -> Self {
        Self { storage_dir }
    }
}

#[async_trait]
impl StorageBackend for Storage {
    async fn delete(&self, path: &str) -> Result<()> {
        fs::remove_file(self.storage_dir.join(path)).await?;
        Ok(())
    }

    async fn get<'a>(&'a self, path: &str) -> Result<BoxStream<'a, Result<Bytes>>> {
        let file = File::open(self.storage_dir.join(path)).await?;
        Ok(ReaderStream::new(file).map_err(Into::into).boxed())
    }

    async fn put(
        &self,
        path: &str,
        input_stream: &mut (dyn Stream<Item = Result<Bytes>> + Send + Unpin),
    ) -> Result<()> {
        let mut file = File::create(self.storage_dir.join(path)).await?;
        while let Some(chunk) = input_stream.next().await.transpose()? {
            file.write_all(&chunk).await?;
        }
        Ok(())
    }
}
