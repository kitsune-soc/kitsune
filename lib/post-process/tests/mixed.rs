use crate::util::parse_to_test_output;
use std::fs;

mod util;

#[test]
fn mixed_content() {
    insta::glob!("input/mixed/mixed_content_*", |path| {
        let post = fs::read_to_string(path).unwrap();
        insta::assert_debug_snapshot!(parse_to_test_output(&post));
    });
}
