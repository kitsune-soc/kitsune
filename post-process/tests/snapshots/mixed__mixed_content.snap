---
source: post-process/tests/mixed.rs
expression: "PostParser::parse(Rule::post, &post).unwrap()"
input_file: post-process/tests/input/mixed/mixed_content_1
---
[
    Pair {
        rule: text,
        span: Span {
            str: "hey,",
            start: 0,
            end: 4,
        },
        inner: [],
    },
    Pair {
        rule: mention,
        span: Span {
            str: " @真島",
            start: 4,
            end: 12,
        },
        inner: [
            Pair {
                rule: component_prefix,
                span: Span {
                    str: " ",
                    start: 4,
                    end: 5,
                },
                inner: [],
            },
            Pair {
                rule: mention_username,
                span: Span {
                    str: "真島",
                    start: 6,
                    end: 12,
                },
                inner: [],
            },
        ],
    },
    Pair {
        rule: text,
        span: Span {
            str: " looking good..",
            start: 12,
            end: 27,
        },
        inner: [],
    },
    Pair {
        rule: hashtag,
        span: Span {
            str: " #龍が如く7",
            start: 27,
            end: 42,
        },
        inner: [
            Pair {
                rule: component_prefix,
                span: Span {
                    str: " ",
                    start: 27,
                    end: 28,
                },
                inner: [],
            },
            Pair {
                rule: hashtag_content,
                span: Span {
                    str: "龍が如く7",
                    start: 29,
                    end: 42,
                },
                inner: [],
            },
        ],
    },
    Pair {
        rule: text,
        span: Span {
            str: " ",
            start: 42,
            end: 43,
        },
        inner: [],
    },
    Pair {
        rule: link,
        span: Span {
            str: "https://upload.wikimedia.org/wikipedia/en/3/3b/GoroMajimafive.jpg",
            start: 43,
            end: 108,
        },
        inner: [
            Pair {
                rule: link_schema,
                span: Span {
                    str: "https",
                    start: 43,
                    end: 48,
                },
                inner: [],
            },
            Pair {
                rule: link_content,
                span: Span {
                    str: "upload.wikimedia.org/wikipedia/en/3/3b/GoroMajimafive.jpg",
                    start: 51,
                    end: 108,
                },
                inner: [],
            },
        ],
    },
]
